data/transactions/logic/eval_test.go

```diff
@@ -0,0 +1,2013 @@
+}
+
+func TestArg(t *testing.T) {
+	t.Parallel()
+	program, err := AssembleString(`arg 0
+arg 1
+==
+arg 2
+arg 3
+!=
+&&
+arg 4
+len
+int 9
+<
+&&`)
+	require.NoError(t, err)
+	cost, err := Check(program, EvalParams{})
+	require.NoError(t, err)
+	require.True(t, cost < 1000)
+	var txn transactions.SignedTxn
+	txn.Lsig.Logic = program
+	txn.Lsig.Args = [][]byte{
+		[]byte("aoeu"),
+		[]byte("aoeu"),
+		[]byte("aoeu2"),
+		[]byte("aoeu3"),
+		[]byte("aoeu4"),
+	}
+	sb := strings.Builder{}
+	pass, err := Eval(program, EvalParams{Trace: &sb, Txn: &txn})
+	if !pass {
+		t.Log(hex.EncodeToString(program))
+		t.Log(sb.String())
+	}
+	require.NoError(t, err)
+	require.True(t, pass)
+}
+
+const globalTestProgram = `global MinTxnFee
+int 123
+==
+global MinBalance
+int 1000000
+==
+&&
+global MaxTxnLife
+int 999
+==
+&&
+global ZeroAddress
+txn CloseRemainderTo
+==
+&&
+global TimeStamp
+int 2069
+==
+&&
+global Round
+int 999999
+==
+&&
+global GroupSize
+int 1
+==
+&&`
+
+func TestGlobal(t *testing.T) {
+	t.Parallel()
+	for _, globalField := range GlobalFieldNames {
+		if !strings.Contains(globalTestProgram, globalField) {
+			t.Errorf("TestGlobal missing field %v", globalField)
+		}
+	}
+	program, err := AssembleString(globalTestProgram)
+	require.NoError(t, err)
+	cost, err := Check(program, EvalParams{})
+	require.NoError(t, err)
+	require.True(t, cost < 1000)
+	var txn transactions.SignedTxn
+	txn.Lsig.Logic = program
+	txgroup := make([]transactions.SignedTxnWithAD, 1)
+	txgroup[0].SignedTxn = txn
+	sb := strings.Builder{}
+	block := bookkeeping.Block{}
+	block.BlockHeader.Round = 999999
+	block.BlockHeader.TimeStamp = 2069
+	proto := config.ConsensusParams{
+		MinTxnFee:       123,
+		MinBalance:      1000000,
+		MaxTxnLife:      999,
+		LogicSigVersion: 1,
+	}
+	ep := EvalParams{
+		Trace:    &sb,
+		Txn:      &txn,
+		Block:    &block,
+		Proto:    &proto,
+		TxnGroup: txgroup,
+	}
+	pass, err := Eval(program, ep)
+	if !pass {
+		t.Log(hex.EncodeToString(program))
+		t.Log(sb.String())
+	}
+	require.NoError(t, err)
+	require.True(t, pass)
+}
+
+const testTxnProgramText = `txn Sender
+arg 0
+==
+txn Receiver
+arg 1
+==
+&&
+txn CloseRemainderTo
+arg 2
+==
+&&
+txn VotePK
+arg 3
+==
+&&
+txn SelectionPK
+arg 4
+==
+&&
+txn Note
+arg 5
+==
+&&
+txn Fee
+int 1337
+==
+&&
+txn FirstValid
+int 42
+==
+&&
+txn LastValid
+int 1066
+==
+&&
+txn Amount
+int 1000000
+==
+&&
+txn VoteFirst
+int 1317
+==
+&&
+txn VoteLast
+int 17776
+==
+&&
+txn VoteKeyDilution
+int 1
+==
+&&
+txn Type
+byte 0x706179
+==
+&&
+txn TypeEnum
+int 1
+==
+&&
+txn XferAsset
+arg 6
+==
+&&
+txn AssetAmount
+int 1234
+==
+&&
+txn AssetSender
+arg 1
+==
+&&
+txn AssetReceiver
+arg 2
+==
+&&
+txn AssetCloseTo
+arg 0
+==
+&&
+txn GroupIndex
+int 3
+==
+&&
+txn TxID
+arg 7
+==
+&&`
+
+func TestTxn(t *testing.T) {
+	t.Parallel()
+	for _, txnField := range TxnFieldNames {
+		if !strings.Contains(testTxnProgramText, txnField) {
+			t.Errorf("TestTxn missing field %v", txnField)
+		}
+	}
```

Comment:
It would be good to have this test in `TestGtxn` as well. (Almost forgot to add a gtxn implementation of something I added.)